### ISSUE TITLE
Issue with zip files uploaded to minio

### ISSUE THREAD

**Issue by milenkovicm** [durch/rust-s3#147]
**Describe the bug**

Looks like the library is trimming ascii 0 character at the end of the zip file when using `put_object_stream_*` methods. 

This problem may be related to filter at https://github.com/durch/rust-s3/blob/af177f4d2bb5d831ae708cfd442ff8e770fc245c/s3/src/utils.rs#L96 thus bytes at the end of file may go missing after upload.

original zip ending:

<img width="500" alt="original" src="https://user-images.githubusercontent.com/956287/101247617-87aaf880-3712-11eb-96fb-7b312d4e92a0.png">

file retrieved from minio:

<img width="487" alt="uploaded" src="https://user-images.githubusercontent.com/956287/101247630-92fe2400-3712-11eb-8824-fa14f4b468ec.png">

it works as expected if `put_objects` with byte slices are used. 

**To Reproduce**

```rust
 let bucket = Bucket::new_with_path_style(&backend.bucket, backend.region, backend.credentials)
        .expect("bucket created");

 bucket.put_object_stream_blocking("original_test.zip", "minio_test.zip").expect("file uploaded");
```
original file to be uploaded: [original_test.zip](https://github.com/durch/rust-s3/files/5647322/original_test.zip)

file retrieved from minio: [minio_test.zip](https://github.com/durch/rust-s3/files/5647320/minio_test.zip)

**Expected behavior**

all bytes accounted in any case

**Environment**

 - Rust version: [e.g. ` 1.48`]
 - lib version [e.g. `0.26.3`]

Thanks for your help,

**Comment by milenkovicm** [comment]
Just few notes regarding `read_chunk` functions:

- Loop should break if `n` is zero, as read may return n < `BUF_LEN` even if not EOF according to the https://doc.rust-lang.org/std/io/trait.Read.html#tymethod.read
- Wouldn’t it make more sense to take slice of `buffer[..n]` instead of current trimming? This way buffer could be allocated just before loop, and reused.

Thanks

**Comment by milenkovicm** [comment]
Test to demonstrate issue and fix referenced.

If satisfied u can cherry-pick it or i can crate PR.

**Comment by durch** [comment]
@milenkovicm Thank you for both the report and the suggested solution, great stuff. Will get to it tomorrow or the day after

**Comment by milenkovicm** [comment]
No problem @durch, let me know if I can help.

Another thing I've noted is related with `etag` calculation.
- Current version on `master` is not deterministic, as in some edge cases it may fail. 
- Version which is using new `read_chunk` is probably deterministic for constant `CHUNK_SIZE` but does not look right depending on a `CHUNK_SIZE` for `md5` calculation. 

To be pedantic, it does not look right using vector to save `md5`s of each chunk and then calculate `md5` of that vector. Why not updating `md5` for each chunk, at least it will take constant memory for that operation?

**Comment by durch** [comment]
Fixes to read_chunk will be pushed to 0.26.4, and 0.27.0-beta4. `etag` is an interesting beast, chunking gymnastics are due to aws special handling of etags for multipart uploads, not sure how that jives with other providers. [This](https://teppen.io/2018/06/23/aws_s3_etags/) was my guide for implementing them, open to improvements :)

**Comment by milenkovicm** [comment]
Thanks @durch, it looks fixed now.
